### PR TITLE
React: Upgrade `react-docgen` to 6.0.x and improve argTypes

### PR DESCRIPTION
**Pull request by shilman**
Extends #22324 

## What I did

**NOTE: This only applies to _VITE_ projects, not Webpack ones yet!!!**

Extends  #22324:
- [x] Update `react-docgen` to 6.0.x
- [x] Add support for fs resolver

Followup:
- [ ] Update webpack too

Also improve argType generation:
- [x] Add experimental support for converting `unions` of `literals` into `enum`, which generate proper controls
- [x] Fix table summary types for array, object, and function

## How to test in your project

To test this out in a React Vite TS project, install the canary version listed at the top of this PR into a project by manually updating its `package.json` to update all the `@storybook/*` packages from this monorepo:

```
    "@storybook/react": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/react-vite": "0.0.0-pr-23825-sha-7875282b",
    ...
```

Once you've installed the packages, edit `.storybook/main.ts` to contain:
```
  typescript: {
    reactDocgen: 'react-docgen',
  },
```

The behavior should be the same as before, but running `storybook dev` should be dramatically faster than before.

If you run into issues, please comment on this PR!

## How to test in a development sandbox

1. Create a development sandbox in this branch, e.g.
```
yarn task --task sandbox --start-from auto --template react-vite/default-ts
```
2. Edit `.storybook/main.ts` to contain:
```
  typescript: {
    reactDocgen: 'react-docgen',
  },
```
3. Add a new file `starbucks-size.ts` in `src/stories` directory:
```ts
export type StarbucksSize = 'tall' | 'grande' | 'venti';
```
4. Update the CLI-generated `Button.tsx` component with a new prop:
```ts
  /**
   * How large should the button be at Starbucks?
   */
  starbucksSize?: StarbucksSize;
```
5. Observe the result in the controls panel:

image

## Checklist

Needs tests:
- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-23825-sha-7875282b`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-23825-sha-7875282b). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-23825-sha-7875282b`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-23825-sha-7875282b) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/fix-react-docgen-enum`](https://github.com/storybookjs/storybook/tree/shilman/fix-react-docgen-enum) |
| **Commit** | [`7875282b`](https://github.com/storybookjs/storybook/commit/7875282ba0dfebc4e5f6a59ead02e08c7e509d3a) |
| **Datetime** | Tue Aug 22 10:30:44 UTC 2023 (`1692700244`) |
| **Workflow run** | [5937622254](https://github.com/storybookjs/storybook/actions/runs/5937622254) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=23825`_
</details>

### CODE REVIEW

File: code/frameworks/react-vite/tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "rootDir": "./src",
     "types": ["node"],
     "resolveJsonModule": true,
-    "strict": true
+    "skipLibCheck": true
```
Comment (yannbf) [inline code/frameworks/react-vite/tsconfig.json:7]: what was this change necessary for?

Comment (shilman) [reply code/frameworks/react-vite/tsconfig.json:7]: Good question. This PR is mostly changes from #22324 and I guess this is one of those.

Comment (danez) [reply code/frameworks/react-vite/tsconfig.json:7]: This was done, because the TS compiler options here are different than the settings in the react-docgen repo. This lead to one of the CI jobs to fail, because the it was validating react-docgen types with the different options (mostly strict stuff). I fixed a lot of it and enabled some more strict options in react-docgen, so maybe it works now.

**Comment by yannbf** [comment]
Hey @shilman I tested this in Mealdrop with `Button.stories.tsx` and the behavior is definitely different between react-docgen-typescript vs using react-docgen. The installed versions are:

`react-docgen`: **6.0.4**
`@joshwooding/vite-plugin-react-docgen-typescript`: **0.2.1**

#### Default behavior(`react-docgen-typescript`):




#### Behavior with `react-docgen`:

image

This is the reproduction code:

#### Button.tsx

```tsx
import * as React from 'react'
type DefaultProps = {
  /**
   * Clear button styles leaving just a text
   */
  clear?: boolean
  round?: boolean
  /**
   * Is the button large?
   */
  large?: boolean
  /**
   * Does the button have an icon?
   */
  icon?: string
  /**
   * Size of the icon
   */
  iconSize?: number
  /**
   * Is the button disabled?
   */
  disabled?: boolean
  /**
   * Does the button have an icon?
   */
  children?: string | React.ReactNode

  starbucksSize: 'tall' | 'grande' | 'venti'
  /**
   * Optional click handler
   */
  onClick?: () => void
}

type ButtonProps = DefaultProps & React.ComponentProps<typeof StyledButton>

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<React.PropsWithChildren<ButtonProps>> = ({
  children,
  large = false,
  clear = false,
  round = false,
  icon,
  iconSize,
}) => {
  return <div>I am a button</div>
}
```

#### Button.stories.tsx
```tsx
export const Default: Story = {
  args: {
    children: 'Hello world',
  }
}
```

**Comment by yannbf** [comment]
Additionally, if you make the code provide even less metadata, like so:

#### Button.tsx

```diff
export const Button: React.FC<React.PropsWithChildren<ButtonProps>> = (
- {
-  children,
-  large = false,
-  clear = false,
-  round = false,
-  icon,
-  iconSize,
-}
) => {
  return <div>I am a button</div>
}
```

#### Button.stories.tsx
```diff
export const Default: Story = {
-  args: {
-    children: 'Hello world',
-  }
}
```

No controls are detected whatsoever

image

While with `react-docgen-typescript`, you get everything like before:

image

**Comment by danez** [comment]
`React.FC` is not yet supported in react-docgen. I will see if I can find some time for this.

**Comment by yannbf** [comment]
Thanks a lot @danez! Are there other limitations when it comes to comparing react-docgen vs react-docgen-typescript we should know about?

Based on your comment, I made a small change in the component and react-docgen provides much better metadata now:

```diff
-export const Button: React.FC<React.PropsWithChildren<ButtonProps>> = ({
+export const Button = ({
  children,
  large = false,
  clear = false,
  round = false,
  icon,
  iconSize,
  ...props
-}) => {
+}: ButtonProps) => {
```

 Only thing is that it doesn't seem to forward the metadata from the inherited props that come from `React.ComponentProps<typeof StyledButton>` (theme, as, forwardedAs)

image


Here's the full code:

<details>
  <summary>Button.tsx</summary>

```tsx
import * as React from 'react'
import styled, { css } from 'styled-components'

const StyledButton = styled.button<{
  clear: boolean
  large: boolean
  withIcon: boolean
  round: boolean
}>(
  () => css`
    outline: none;
  `
)

type DefaultProps = {
  /**
   * Clear button styles leaving just a text
   */
  clear?: boolean
  round?: boolean
  /**
   * Is the button large?
   */
  large?: boolean
  /**
   * Does the button have an icon?
   */
  icon?: string
  /**
   * Size of the icon
   */
  iconSize?: number
  /**
   * Is the button disabled?
   */
  disabled?: boolean
  /**
   * Does the button have an icon?
   */
  children?: string | React.ReactNode

  starbucksSize: 'tall' | 'grande' | 'venti'
  /**
   * Optional click handler
   */
  onClick?: () => void
}

type ButtonProps = DefaultProps & React.ComponentProps<typeof StyledButton>

/**
 * Primary UI component for user interaction
 */
export const Button = ({
  children,
  large = false,
  clear = false,
  round = false,
  icon,
  iconSize,
  ...props
}: ButtonProps) => {
  return (
    <StyledButton
      type="button"
      large={large}
      clear={clear}
      round={round}
      withIcon={!!icon}
      {...props}
    >
      {icon && <div>Icon here</div>}
      {children}
    </StyledButton>
  )
}

```

</details>

**Comment by shilman** [comment]
@yannbf I wrote about that limitation here several years ago but it's been so long that I completely forgot about it https://github.com/storybookjs/storybook/blob/next/code/addons/docs/react/README.md?plain=1#L155-L169

@danez That would be amazing if you could fix the `React.FC` problem. That's probably a dealbreaker blocking us from making `react-dogen` the default option in Storybook 8.0.

My plan is to get this PR into a mergeable state and release this upgrade as part of 7.5. That will help get it into more peoples' hands and we can see what issues come out of it.

I've also created a follow-on PR here https://github.com/storybookjs/storybook/pull/24165 to make `react-docgen` the default, and will release that in 8.0-alpha. Hopefully that will generate more good feedback and help us gain confidence on the change.

**Review by ndelangen — APPROVED** [review]

**Comment by yannbf** [comment]
> `React.FC` is not yet supported in react-docgen. I will see if I can find some time for this.

Hey @danez did you have any luck with React.FC? Thank you so much for taking a look into it! <3

**Comment by shilman** [comment]
@danez FYI we're going to evaluate https://www.npmjs.com/package/typescript-react-function-component-props-handler and will report back. If it works well, it could be a good candidate to incorporate in the default settings.

**Comment by danez** [comment]
Nice. React.FC is already supported on the main branch of react-docgen but I noticed some other issues that i want to fix first, especially because i had to make a breaking change again.

**Comment by valentinpalkovic** [comment]
@shilman, @danez I took a quick look at the [custom resolver ](https://www.npmjs.com/package/typescript-react-function-component-props-handler), and it failed pretty quickly

Bildschirmfoto 2023-10-05 um 15 55 10


@danez Do you need some support? I would say I don't spend time fixing the third-party resolve, but instead, I would rather wait and/or support your work on the main branch.

**Comment by shilman** [comment]
@danez that's great news. please let us know how we can support! we're going to release this upgrade PR in Storybook 7.5 next week, and will merge the companion PR https://github.com/storybookjs/storybook/pull/24165 for 8.0 once FC support is merged & released. Happy to adjust the plans if it helps you out in any way.

**Comment by rachelholtz** [comment]
I went to try this out but I am getting an error when running storybook.
I updated all @storybook packages:
```
    "@storybook/addon-a11y": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/addon-essentials": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/addon-interactions": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/addon-links": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/blocks": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/react": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/react-vite": "0.0.0-pr-23825-sha-7875282b",
```
    
Already had  `typescript: {reactDocgen: 'react-docgen'}` in main.ts.

Did a clean install.

When I run storybook I get this error:

@storybook/cli v7.4.3
WARN   Failed to load preset: "@storybook/react-vite/preset"
Error: Cannot find module '@storybook/react-vite/preset'
Error: Invariant failed: No builder configured in core.builder

So I tried to specifically install the canary of @storybook/cli and it didn't help.

I have reread the instructions above and I feel like I am missing something. Storybook runs fine with 7.1.0 but was hoping to test the updates for react-docgen. We have issues around pulling in descriptions for arg types when a component Type is extending another component's types.

**Comment by ndelangen** [comment]
cc @shilman ☝️

**Comment by harryyaprakov** [comment]
First, great work on this excellent product!
Second, do you guys plan on migrating to v7.0 of docgen as released any time soon?

I tried running v7.0 (by npm override) with sb v7.5 and am getting unexpected and strange errors.

**Comment by yannbf** [comment]
Hey @harryyaprakov thank you! We are planning on doing that for Storybook 7.6. Can you maybe elaborate and share a reproduction that can display the errors you are mentioning? It will be really useful for us.

cc @shilman
